File: test_week6.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const vm = require('vm');
const path = require('path');

const root = path.resolve(__dirname);
function load(file, context) {
  vm.runInContext(fs.readFileSync(path.join(root, file), 'utf8'), context, { filename: file });
}
function assert(condition, message) {
  if (!condition) throw new Error(message);
}

// Contract test for the deterministic explanation layer.
const explainContext = vm.createContext({
  console,
  Number,
  String,
  Promise,
  isFinite,
  NexlunaUnits: {
    length: {
      label: 'الطول',
      units: [['m', 'متر'], ['ft', 'قدم']],
      provenance: { source: 'test-fixture' },
      factors: { m: 1, ft: 0.3048 }
    }
  },
  NexMCP: {
    convert(category, from, to, value) {
      assert(category === 'length' && from === 'm' && to === 'ft', 'unexpected deterministic request');
      return Number(value) / 0.3048;
    },
    provenanceFor() { return { source: 'test-fixture' }; }
  }
});
explainContext.window = explainContext;
load('assets/js/explain.js', explainContext);
(async () => {
  const baseline = explainContext.NexExplain.deterministic('length', 'm', 'ft', 1);
  assert(baseline.output.value > 3.2808 && baseline.output.value < 3.2809, 'deterministic explanation value is wrong');
  assert(baseline.calculation.deterministic === true, 'baseline is not marked deterministic');
  assert(baseline.calculation.network_required === false, 'baseline unexpectedly requires network');
  const fallback = await explainContext.NexExplain.explain({ category: 'length', from: 'm', to: 'ft', value: 1 });
  assert(fallback.source === 'deterministic-fallback', 'missing-model fallback contract failed');
  const verified = explainContext.NexExplain.verifyModelResult({ value: baseline.output.value, explanation: 'verified' }, baseline);
  assert(verified && verified.source === 'optional-model-verified', 'verified model result contract failed');
  const rejected = explainContext.NexExplain.verifyModelResult({ value: 999, explanation: 'wrong' }, baseline);
  assert(rejected === null, 'wrong model value was accepted');

  // Lightweight DOM fixture for embed.js, including URL prefill and postMessage.
  function element() {
    return {
      value: '', textContent: '', innerHTML: '', dataset: {}, options: [], listeners: {},
      appendChild(child) {
        this.options.push(child);
        if (child.selected || (this.value === '' && this.options.length === 1)) this.value = child.value;
      },
      addEventListener(type, fn) { this.listeners[type] = fn; }
    };
  }
  const elements = {
    '#embed-category': element(), '#embed-from': element(), '#embed-to': element(),
    '#embed-value': element(), '#embed-result': element(), '#embed-form': element()
  };
  const messages = [];
  const embedContext = vm.createContext({
    console,
    Number,
    String,
    URLSearchParams,
    window: null,
    document: {
      querySelector(selector) { return elements[selector]; },
      createElement() { return { value: '', textContent: '', selected: false }; }
    },
    NexlunaUnits: {
      length: { label: 'الطول', units: [['m', 'متر'], ['ft', 'قدم']] }
    },
    NexMCP: {
      convert(category, from, to, value) {
        assert(category === 'length' && from === 'm' && to === 'ft', 'embed request was not prefilled');
        return Number(value) / 0.3048;
      }
    }
  });
  embedContext.window = embedContext;
  embedContext.location = { search: '?category=length&from=m&to=ft&v=2' };
  embedContext.parent = { postMessage(message, origin) { messages.push({ message, origin }); } };
  load('assets/js/embed.js', embedContext);
  assert(elements['#embed-category'].value === 'length', 'embed category prefill failed');
  assert(elements['#embed-from'].value === 'm' && elements['#embed-to'].value === 'ft', 'embed unit prefill failed');
  assert(elements['#embed-value'].value === '2', 'embed value prefill failed');
  assert(elements['#embed-result'].dataset.state === 'success', 'embed initial conversion did not succeed');
  assert(messages.length === 1, 'embed did not emit exactly one result message');
  assert(messages[0].message.type === 'nexluna-result', 'wrong postMessage type');
  assert(messages[0].message.result > 6.56 && messages[0].message.result < 6.57, 'wrong embed result');
  assert(messages[0].origin === '*', 'unexpected postMessage origin contract');

  console.log('PASS: test_week6.js');
})().catch((error) => {
  console.error('FAIL: test_week6.js:', error.message);
  process.exitCode = 1;
});

File: test_week6.py
#!/usr/bin/env python3
"""Week 6 integration gate: explanation, embed contract, generation, and data systems."""
from pathlib import Path
import re
import subprocess
import sys

ROOT = Path(__file__).resolve().parent


def require(condition, message):
    if not condition:
        raise AssertionError(message)


def script_order(html, scripts):
    positions = [html.find(f'src="/assets/js/{name}"') for name in scripts]
    require(all(position >= 0 for position in positions), f"missing script in generated page: {scripts}")
    require(positions == sorted(positions), f"script order is invalid: {scripts}")


def main():
    # Generated pages must expose the deterministic explanation stack in order.
    home = (ROOT / "index.html").read_text(encoding="utf-8")
    category = (ROOT / "converters" / "length.html").read_text(encoding="utf-8")
    pair = (ROOT / "convert" / "km-to-m.html").read_text(encoding="utf-8")
    script_order(home, ["units.generated.js", "explain.js", "webmcp.js", "converter.js"])
    script_order(category, ["units.generated.js", "explain.js", "webmcp.js", "converter.js"])
    script_order(pair, ["units.generated.js", "explain.js", "webmcp.js"])

    embed = (ROOT / "embed.html").read_text(encoding="utf-8")
    for element_id in ("embed-category", "embed-from", "embed-to", "embed-value", "embed-result", "embed-form"):
        require(f'id="{element_id}"' in embed, f"embed element missing: {element_id}")
    require("/assets/js/units.generated.js" in embed, "embed must load canonical units")
    require("/assets/js/webmcp.js" in embed, "embed must load WebMCP")
    require("/assets/js/embed.js" in embed, "embed must load embed controller")

    converter = (ROOT / "assets" / "js" / "converter.js").read_text(encoding="utf-8")
    require("explainBtn" in converter and "NexExplain.explain" in converter, "converter explanation control is not wired")
    require("nx-data-system" in converter and "KiB" in converter and "KB" in converter, "data system toggle is not wired")
    require("ic('spark')" in converter, "explanation button icon is missing")

    explain = (ROOT / "assets" / "js" / "explain.js").read_text(encoding="utf-8")
    require("deterministic-fallback" in explain and "optional-model-verified" in explain, "explanation fallback contract is incomplete")
    require("network_required: false" in explain, "explanation must declare no network requirement for baseline")

    sw = (ROOT / "sw.js").read_text(encoding="utf-8")
    for asset in ("/assets/js/explain.js", "/assets/js/embed.js", "/embed.html"):
        require(asset in sw, f"service worker missing Week 6 asset: {asset}")

    subprocess.run(["node", str(ROOT / "test_week6.js")], cwd=ROOT, check=True)
    print("PASS: Week 6 integration gate")


if __name__ == "__main__":
    try:
        main()
    except (AssertionError, subprocess.CalledProcessError) as exc:
        print(f"FAIL: {exc}", file=sys.stderr)
        raise SystemExit(1)
